core/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java

```diff
@@ -125,6 +132,9 @@ public BulkRequest add(IndexRequest request, @Nullable Object payload) {
     }
 
     BulkRequest internalAdd(IndexRequest request, @Nullable Object payload) {
+        if (request.refresh()) {
```

Comment:
we need to do this in the validate method. The list of requests is exposed and people can change them after adding.